### DISCUSSION TITLE
Strange BIOS Boot Option for "Whole Disks"

### DISCUSSION THREAD

**Discussion by MarioG-X** [dakanji/RefindPlus#130] [General]
Head scratcher: When booting from a USB stick that has multiple Linux systems in BTRFS, and a Gparted partition. The menu adds an option at the end of the display list on the screen to boot the whole disk in BIOS, of course this does not work. The log shows it found legacy boot code but I looked at the EFI partition and the gparted partition and found nothing. I ran gparted and it shows as GPT only, not a hybrid GPT/MBR table which RP shows. I cannot find anything in the log to direct me to not display that entry. I don't think this is a bug but something I am missing. The log is attached.
[22x07w4315.log](https://github.com/dakanji/RefindPlus/files/10180364/22x07w4315.log)

**Comment by dakanji** [comment]
Not quite sure. I just press the '-' key on such unwanted items.

Will investigate these at some point

**Comment by dakanji** [comment]
Just out of interest, do you mean to say the behaviour in RefindPlus is different to rEFInd on this?

As an aside, you shouldn't need to load an ISO-9660 driver on a Mac as this has been part of the firmware since the `New World ROM`
https://apple.fandom.com/wiki/New_World_ROM

**Comment by startergo** [comment]
I think this nuisance has always been there. 
as far as I remember rEFInd code looks for the magic number to define legacy boot it does not look at the partitioning. Remember Apple partitions the disk with a hybrid MBR scheme which is a combination of both GPT and MBR.  @joevt what do you think?

**Comment by MarioG-X** [comment]
While working on this I created #131 which may or may not be related. I have not been able to find anything that shows a hybrid table for the internal disks, only shows as GPT, and RP AA did not, AB/AC does.

I removed the ISO driver.  Thanks. I will test Refind, and look into linux utilities.

**Comment by MarioG-X** [comment]
BTW, the reason for this is that I was looking for things to speed up RP time when booting the USB which has multiple Linux's and gparted.

I tested Refind and it displays that same erroneous entry on the screen.

Linux parted, sfdisk, and fdisk showed as GPT. However, gdisk has a line "MBR: Protective". So it may be a macOS thing, but strange Refind sees it as a bootable partition. If Refind/RP does not find a module to boot, should it not list it?

**Comment by dakanji** [comment]
Protective MBR is not an issue and basically standard.

The issue is as described. rEFIt variants id disk types by looking for "magic" values which should only be present for that type but is apparently not always the case.

**Comment by joevt** [comment]
I would use `dd` and `xxd` on the first block of the disk to see if it has boot code. iPartition.app has an option to remove and add boot code.

I made [dumpvols.sh](https://gist.github.com/joevt/a99e3af71343d8242e0078ab4af39b6c) to look at problems like this while I was playing with Windows XP, Windows Vista, and Windows 7, and FreeDOS partitions on my Mac Pro 2008. Then I replaced Windows XP and Vista with Windows 8 and 10. My Linux partitions boot using EFI.

**Comment by dakanji** [comment]
Took a quick look and it seems at first glance that EXFAT formatted disks that do not contain any bootable legacy volumes are likely to end up being considered as being bootable.

**Reply by joevt** [reply]
EXFAT is a format for a partition - nothing to do with the MBR.

Maybe during one of the linux installs, some boot code was added to the MBR for legacy BIOS boot compatibility?

Didn't one of the rEFInd tools have an option to add boot code?

The MBR should be mostly zeros (no boot code at offset 0) except for the partition table and the `55aa` at the end. For a GPT disk, there should be only 1 MBR partition with type `EE` that extends from block 1 to the end of the disk.

```
joevt@Joes-Mac-Pro ~ % sudo dd if=/dev/disk0 count=1 2>/dev/null | xxd
Password:
00000000: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000010: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000020: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000030: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000040: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000050: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000060: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000070: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000080: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000090: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000100: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000110: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000120: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000130: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000140: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000150: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000160: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000170: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000180: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000190: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001b0: 0000 0000 0000 0000 0000 0000 0000 00ff  ................
000001c0: ffff eeff ffff 0100 0000 af1a c86f 0000  .............o..
000001d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001f0: 0000 0000 0000 0000 0000 0000 0000 55aa  ..............U.

joevt@Joes-Mac-Pro ~ % sudo fdisk /dev/disk0
Disk: /dev/disk0	geometry: 116737/255/63 [1875385008 sectors]
Signature: 0xAA55
         Starting       Ending
 #: id  cyl  hd sec -  cyl  hd sec [     start -       size]
------------------------------------------------------------------------
 1: EE 1023 255  63 - 1023 255  63 [         1 - 1875385007] <Unknown ID>
 2: 00    0   0   0 -    0   0   0 [         0 -          0] unused      
 3: 00    0   0   0 -    0   0   0 [         0 -          0] unused      
 4: 00    0   0   0 -    0   0   0 [         0 -          0] unused      

joevt@Joes-Mac-Pro ~ % sudo gpt -r show /dev/disk0
       start        size  index  contents
           0           1         PMBR
           1           1         Pri GPT header
           2          32         Pri GPT table
          34           6         
          40      409600      1  GPT part - C12A7328-F81F-11D2-BA4B-00A0C93EC93B
      409640   115343360      2  GPT part - 48465300-0000-11AA-AA11-00306543ECAC
   115753000     1269536      3  GPT part - 426F6F74-0000-11AA-AA11-00306543ECAC
   117022536   126953120      4  GPT part - 48465300-0000-11AA-AA11-00306543ECAC
   243975656     1269536      5  GPT part - 426F6F74-0000-11AA-AA11-00306543ECAC
   245245192   265878576         
   511123768   264087376      6  GPT part - 7C3457EF-0000-11AA-AA11-00306543ECAC
   775211144   487298304      7  GPT part - 7C3457EF-0000-11AA-AA11-00306543ECAC
  1262509448   390625000      8  GPT part - 7C3457EF-0000-11AA-AA11-00306543ECAC
  1653134448   222045584         
  1875180032      102400      9  GPT part - 48465300-0000-11AA-AA11-00306543ECAC
  1875282432      102536     10  GPT part - 48465300-0000-11AA-AA11-00306543ECAC
  1875384968           7         
  1875384975          32         Sec GPT table
  1875385007           1         Sec GPT header
```

**Comment by dakanji** [comment]
> EXFAT is a format for a partition - nothing to do with the MBR

Indeed but it is being taken into consideration here on Macs:
https://github.com/dakanji/rEFInd/blob/master/refind/lib.c#L769-L773
This is before the OS Specific checks and overrides such.

Later on, for Macs and others, it checks whether a Whole Disk Volume has any legacy items and hides it if so. 
If it doesn't, it displays the Whole Disk item as a boot option:
https://github.com/dakanji/rEFInd/blob/0d46f46ed7a2d453c9320469f211c5348120a1cd/refind/launch_legacy.c#L663-L678

I got it wrong at the first look though ... the attempt is to exclude EXFAT:
https://github.com/dakanji/rEFInd/commit/f06fd828a5e1a814986707a9f3b233553995d5f2

So, if this matches on a Mac and the disk is not EXFAT and it later turns out the Whole Disk does not have loaders, it assumes the Whole Disk item is bootable:
`(*((UINT16 *)(Buffer + 510)) == 0xaa55 && Buffer[0] != 0)`

Presumably that is what is happening here ... potentially due to this:
> Maybe during one of the linux installs, some boot code was added to the MBR for legacy BIOS boot compatibility?

**Reply by startergo** [reply]
This assumption by the code seems incorrect though don’t you think? The code should be looking for Microsoft stuff like NTLDR or BOOTMGR to declare it MBR bootable. Are there Linux flavors booting in MBR mode?

**Reply by dakanji** [reply]
Read the thread to the bottom.

BTW, just stumbled on this ... Might be better to always post at the bottom

**Comment by MarioG-X** [comment]
I executed commands on the USB and, however disk0 will not permit operation, but got disk1 to compare to disk5 (USB) which is the problem. The extra entry only shows up when using the USB stick. The USB stick block 0 is nothing that is expected. Disk1 is much closer to yours. Something is strange on the USB stick I guess. BTW, I have 2 of USBs identical so they do the same thing.

I wonder if installing GParted could have caused this? Also wonder if I can just copy block 0 of disk1 to disk5 to fix it?

Here is my output for the USB stick:
```
mario@FS-Monterey local % sudo dd if=/dev/disk5 count=1 2>/dev/null | xxd
00000000: 33c0 fa8e d88e d0bc 007c 89e6 0657 8ec0  3........|...W..
00000010: fbfc bf00 06b9 0001 f3a5 ea1f 0600 0052  ...............R
00000020: 52b4 41bb aa55 31c9 30f6 f9cd 1372 1381  R.A..U1.0....r..
00000030: fb55 aa75 0dd1 e973 0966 c706 8d06 b442  .U.u...s.f.....B
00000040: eb15 5ab4 08cd 1383 e13f 510f b6c6 40f7  ..Z......?Q...@.
00000050: e152 5066 31c0 6699 e866 00e8 3501 4d69  .RPf1.f..f..5.Mi
00000060: 7373 696e 6720 6f70 6572 6174 696e 6720  ssing operating 
00000070: 7379 7374 656d 2e0d 0a66 6066 31d2 bb00  system...f`f1...
00000080: 7c66 5266 5006 536a 016a 1089 e666 f736  |fRfP.Sj.j...f.6
00000090: f47b c0e4 0688 e188 c592 f636 f87b 88c6  .{.........6.{..
000000a0: 08e1 41b8 0102 8a16 fa7b cd13 8d64 1066  ..A......{...d.f
000000b0: 61c3 e8c4 ffbe be7d bfbe 07b9 2000 f3a5  a......}.... ...
000000c0: c366 6089 e5bb be07 b904 0031 c053 51f6  .f`........1.SQ.
000000d0: 0780 7403 4089 de83 c310 e2f3 4874 5b79  ..t.@.......Ht[y
000000e0: 3959 5b8a 4704 3c0f 7406 247f 3c05 7522  9Y[.G.<.t.$.<.u"
000000f0: 668b 4708 668b 5614 6601 d066 21d2 7503  f.G.f.V.f..f!.u.
00000100: 6689 c2e8 acff 7203 e8b6 ff66 8b46 1ce8  f.....r....f.F..
00000110: a0ff 83c3 10e2 cc66 61c3 e876 004d 756c  .......fa..v.Mul
00000120: 7469 706c 6520 6163 7469 7665 2070 6172  tiple active par
00000130: 7469 7469 6f6e 732e 0d0a 668b 4408 6603  titions...f.D.f.
00000140: 461c 6689 4408 e830 ff72 2766 813e 007c  F.f.D..0.r'f.>.|
00000150: 5846 5342 7509 6683 c004 e81c ff72 1381  XFSBu.f......r..
00000160: 3efe 7d55 aa0f 85f2 febc fa7b 5a5f 07fa  >.}U.......{Z_..
00000170: ffe4 e81e 004f 7065 7261 7469 6e67 2073  .....Operating s
00000180: 7973 7465 6d20 6c6f 6164 2065 7272 6f72  ystem load error
00000190: 2e0d 0a5e acb4 0e8a 3e62 04b3 07cd 103c  ...^....>b.....<
000001a0: 0a75 f1cd 18f4 ebfd 0000 0000 0000 0000  .u..............
000001b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001c0: 0200 eefe ffff 0100 0000 ffff cf1d 0000  ................
000001d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001f0: 0000 0000 0000 0000 0000 0000 0000 55aa  ..............U.
```

Compare that to an internal drive:
```
mario@FS-Monterey local % sudo dd if=/dev/disk1 count=1 2>/dev/null | xxd
00000000: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000010: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000020: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000030: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000040: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000050: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000060: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000070: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000080: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000090: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000100: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000110: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000120: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000130: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000140: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000150: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000160: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000170: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000180: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000190: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001c0: 0200 eeff ffff 0100 0000 af88 e0e8 0000  ................
000001d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000001f0: 0000 0000 0000 0000 0000 0000 0000 55aa  ..............U.
```

```
mario@FS-Monterey local % sudo fdisk /dev/disk5                    
Disk: /dev/disk5	geometry: 31134/255/63 [500170752 sectors]
Signature: 0xAA55
         Starting       Ending
 #: id  cyl  hd sec -  cyl  hd sec [     start -       size]
------------------------------------------------------------------------
 1: EE    0   0   2 - 1023 254  63 [         1 -  500170751] <Unknown ID>
 2: 00    0   0   0 -    0   0   0 [         0 -          0] unused      
 3: 00    0   0   0 -    0   0   0 [         0 -          0] unused      
 4: 00    0   0   0 -    0   0   0 [         0 -          0] unused      
```

```
mario@FS-Monterey local % sudo gpt -r show /dev/disk5                    
      start       size  index  contents
          0          1         PMBR
          1          1         Pri GPT header
          2         32         Pri GPT table
         34       2014         
       2048     409600      1  GPT part - C12A7328-F81F-11D2-BA4B-00A0C93EC93B
     411648   16777216      2  GPT part - 0657FD6D-A4AB-43C4-84E5-0933C84B4F4F
   17188864   67108864      3  GPT part - 0FC63DAF-8483-4772-8E79-3D69D8477DE4
   84297728   67108864      4  GPT part - 0FC63DAF-8483-4772-8E79-3D69D8477DE4
  151406592   67108864      5  GPT part - 0FC63DAF-8483-4772-8E79-3D69D8477DE4
  218515456   67108864      6  GPT part - 0FC63DAF-8483-4772-8E79-3D69D8477DE4
  285624320   67108864      7  GPT part - 0FC63DAF-8483-4772-8E79-3D69D8477DE4
  352733184   67108864      8  GPT part - 0FC63DAF-8483-4772-8E79-3D69D8477DE4
  419842048    8388608      9  GPT part - EBD0A0A2-B9E5-4433-87C0-68B6B72699C7
  428230656   71940063         
  500170719         32         Sec GPT table
  500170751          1         Sec GPT header
```

**Comment by MarioG-X** [comment]
Problem solved by : sudo dd if=/dev/disk1 count=1 of=/dev/disk5

When I created the USB, the OpenSUSE, Tumbleweed, and ElementaryOS partitions were created then loaded via rsync. But I don't remember how I created the GParted partition and think that was the cause of the issue, probably loading that code into block 0. That useless entry no longer shows up in RP menu. Thanks for all your help guiding me to a solution!

Also, it seems since something was there RP was correct displaying the menu. I think it did not work/boot because it was BIOS code which won't run on a Mac.

**Reply by startergo** [reply]
No, MBR boot is the default boot of Bootcamp drives on cMP. It is actually UEFI boot which is not supported nor designed by Apple to run on these machines. The problem with the rEFIt code not identifying properly an MBR bootable disk is still valid. it is not enough for the code to look only at the magic byte presence, but should also include other parameters.

**Comment by dakanji** [comment]
> it is not enough for the code to look only at the magic byte presence

That is not just what is done though. There are hundreds of lines of code dedicated to determining whether an MBR table exists and whether/how this is populated.

Everytime a disk is "wrongly" presented, it is because something has filled the MBR table with boot code. A lot of what it done in the code is filtering spurious cases out.

**Comment by dakanji** [comment]
> A lot of what it done in the code is filtering spurious cases out.

In this particular case, it is possible to add something to filter it out by excluding stuff that contain strings it contains such as `Operating system load error` or `Missing operating system` but this will need first checking that it will not disable something else.

All in all, it seems it already does a reasonable job of determining what is really bootable and for those that have written stuff that can't be booted and are not covered, it is a simple one-time action to click the `minus` key on the entry to remove them.

Upstream might have more capacity to check and update the filters however.

**Comment by joevt** [comment]
NTLDR or BOOTMGR are things that the code of the VBR of a Windows partition would look for.

Anything can put boot code in the [MBR](https://en.wikipedia.org/wiki/Master_boot_record).

grub for legacy BIOS boot of linux has boot code for the MBR.

The boot code in the MBR (whether it's installed by Windows or Linux) usually just jumps to the boot code that's in the first block of a partition ([VBR](https://en.wikipedia.org/wiki/Volume_boot_record)) that is marked as active in the MBR. The boot code can do anything it wants - it could ignore the partition table and do something else - which means an MBR doesn't need to have a partition table. Since the boot code can do anything, rEFInd can't determine what the boot code needs to work, so rEFInd needs to assume that it does work and will boot something if the Mac or PC supports legacy BIOS boot. There should be no filters on the contents of the boot code - if it's not zero then it's code.

Older Macs support legacy BIOS boot via a BootCamp executable in EFI which rEFInd knows about and can look for. These older Macs can boot Linux with EFI since Linux knows how to work with the old EFI version. Extra work is needed for these old Macs to boot Windows using UEFI but people have done it (SSDT patching via something like Open Core is required for some things like audio).

Newer Macs don't have the BootCamp executable so rEFInd shouldn't show a legacy BIOS boot in that case. These newer Macs use UEFI to boot Linux and Windows.

PCs support legacy BIOS boot via a standard UEFI method which rEFInd knows about.

rEFInd may look at the boot code of a VBR to determine what kind of OS is installed to a partition. If there's a file system driver for the partition then rEFInd can look for certain files to determine the OS.

rEFInd can either legacy BIOS boot a partition (if the MBR and VBR has boot code) or it can EFI boot a partition. For legacy BIOS boot of a selected partition, rEFInd does have an option to allow changing the active partition of an MBR.

**Comment by joevt** [comment]
When doing partitioning, I usually just do it in macOS.
I use dumpvols.sh before partitioning or installing an OS to get a snapshot of the MBRs and partitions and VBRs. Then I can do a second dump after completing the job to see if anything got changed unexpectedly (such as the active partition or the boot code or the partition tables in the MBR and GPT)

**Comment by dakanji** [comment]
> NTLDR or BOOTMGR are things that the code of the VBR of a Windows partition would look for.

Windows is easy and already covered:
https://github.com/dakanji/rEFInd/blob/0d46f46ed7a2d453c9320469f211c5348120a1cd/refind/lib.c#L817-L829

> Since the boot code can do anything, rEFInd needs to assume that it does work and will boot something if the Mac or PC supports legacy BIOS boot.

Indeed.

> There should be no filters on the contents of the boot code - if it's not zero then it's code.

Correct, but some filters, such as the `EXFAT` one from earlier and a few such as these below, have been added for practicality:
https://github.com/dakanji/rEFInd/blob/0d46f46ed7a2d453c9320469f211c5348120a1cd/refind/lib.c#L866-L878

It does comes with the risk of disabling something else and should really not be done indeed.
Seems not to cause issues so far but the possibility is there.
